include/RAJA/policy/cuda/raja_cudaerrchk.hpp

```diff
@@ -25,19 +25,193 @@
 #if defined(RAJA_ENABLE_CUDA)
 
 #include <iostream>
+#include <utility>
+#include <algorithm>
+#include <tuple>
+#include <array>
+#include <string_view>
 #include <string>
+#include <sstream>
+#include <stdexcept>
 
 #include <cuda.h>
 #include <cuda_runtime.h>
 
+#include "RAJA/util/Printing.hpp"
 #include "RAJA/util/macros.hpp"
+#include "RAJA/util/for_each.hpp"
+
+#include "cub/util_type.cuh"
 
 namespace RAJA
 {
 
+namespace detail
+{
+
+template < >
+struct StreamInsertHelper<CUDA_DIM_T&>
+{
+  CUDA_DIM_T& m_val;
+
+  std::ostream& operator()(std::ostream& str) const
+  {
+    return str << "{" << m_val.x
+               << "," << m_val.y
+               << "," << m_val.z
+               << "}";
+  }
+};
+///
+template < >
+struct StreamInsertHelper<CUDA_DIM_T const&>
+{
+  CUDA_DIM_T const& m_val;
+
+  std::ostream& operator()(std::ostream& str) const
+  {
+    return str << "{" << m_val.x
+               << "," << m_val.y
+               << "," << m_val.z
+               << "}";
+  }
+};
+
+template < typename R >
+struct StreamInsertHelper<::cub::DoubleBuffer<R>&>
+{
+  ::cub::DoubleBuffer<R>& m_val;
+
+  std::ostream& operator()(std::ostream& str) const
+  {
+    return str << "{" << m_val.Current() << "," << m_val.Alternate() << "}";
+  }
+};
+///
+template < typename R >
+struct StreamInsertHelper<::cub::DoubleBuffer<R> const&>
+{
+  ::cub::DoubleBuffer<R> const& m_val;
+
+  std::ostream& operator()(std::ostream& str) const
+  {
+    return str << "{" << m_val.Current() << "," << m_val.Alternate() << "}";
+  }
+};
+
+///
+///////////////////////////////////////////////////////////////////////
+///
+/// Utility assert method used in CUDA operations to report CUDA
+/// error codes when encountered.
+///
+///////////////////////////////////////////////////////////////////////
+///
+#define RAJA_INTERNAL_CUDA_CHECK_API_CALL(func, ...)                      \
+  do {                                                                    \
+    cudaError_t code_56792578 = func(__VA_ARGS__);                        \
+    if (code_56792578 != cudaSuccess)                                     \
```

Comment:
Maybe add a comment to add `[[unlikely]]` when we support c++20.

Reply:
done